### ISSUE TITLE
Normalize API surface to use abstractions vs concrete implementations

### ISSUE THREAD

**Issue by egil** [atc-net/atc-rest-api-generator#93]
The generated code should follow the general API design pattern related to MA0016.

Related: https://github.com/meziantou/Meziantou.Analyzer/blob/main/docs/Rules/MA0016.md

**Comment by egil** [comment]
@LarsSkovslund / @rickykaare / @davidkallesen / @perkops / @cjakobsen / et. al. please share your input on the suggested changes above. 

Here are a few questions/thoughts:

1. Is this all the places where we are using concrete types?
2. I like to keep parameters as `IReadOnlyList<T>` since there are very few scenarios where the endpoint handler needs to modify the received list, and in those cases, a LINQ statement or similar will usually result in a new array/list being created anyway.
3. Should models also default to `IReadOnlyList<T>` or perhaps be `ICollection<T>/IList<T>`? I do not see a point in making the returned collections mutable, since they are just getting serialized to JSON, and having the type as an `IReadOnlyList<T>` means pretty much any collection type in the framework can be used to initialize the property. But maybe I am missing something here?

**Comment by cjakobsen** [comment]
I support the suggested changes, it will improve usability of the generated code and add more flexibility in the generator.

Why change the default value to Array.Empty()?

I noticed that in Atc.Rest.Results.ResultFactory we use string[], changing this to IEnumerable<string> could be another improvement.

For the non-breaking part I suggest to just move ahead. For the breaking part we need to define the release process for this, including how this is communicated and versioned.

**Comment by LarsSkovslund** [comment]
@cjakobsen  because `Array.Empty<T>` is more performant than `new string[]`.

**Comment by cjakobsen** [comment]
That is correct but if we want to optimize for that as well then we should be even more crisp and use `Enumerable.Empty<T>()`

**Comment by egil** [comment]
> That is correct but if we want to optimize for that as well then we should be even more crisp and use `Enumerable.Empty<T>()`

Interesting, wasn't aware of that. Looking at the code involved, I am not sure it is more "crisp" though:

[Array.Empty<T>()](https://source.dot.net/#System.Private.CoreLib/Array.cs,710) is just an `static new T[0]` under the hood.

[Enumerable.Empty<T>()](https://source.dot.net/#System.Linq/System/Linq/Enumerable.SpeedOpt.cs,10) points to a [EmptyPartition<TElement>()](https://source.dot.net/#System.Linq/System/Linq/Partition.SpeedOpt.cs,20) that does seem to be a bit more complex.

What are your reasons to prefer `Enumerable.Empty<T>()` Claus?
